hpy/devel/__init__.py

```diff
@@ -295,6 +296,12 @@ def _finalize_hpy_ext(self, ext):
         ext.include_dirs += self.hpydevel.get_extra_include_dirs()
         ext.sources += self.hpydevel.get_extra_sources()
         ext.define_macros.append(('HPY', None))
+        if self.compiler == 'msvc' or get_default_compiler() == 'msvc':
+            # If using MSVC, we need to enable the new preprocessor otherwise
+            # our variable macros (in particular HPyType_HELPERS) won't
+            # properly work. See also:
+            # https://learn.microsoft.com/en-us/cpp/preprocessor/preprocessor-experimental-overview?view=msvc-170
+            ext.extra_compile_args.append('/Zc:preprocessor')
```

Comment:
@antocuni @hodgestar Before I merge, I ask you to have a look on this.
The MSVC preprocessor isn't conforming (by default) to the C standard. This hurts us in the variadic macro `HPyType_HELPERS`. Consider this example:
```
#define BAR(A, B, ...) A
#define FOO(...) BAR(__VA_ARGS__, "default")

FOO("hello", "world")
```
Result on MSVC: `"hello", "world"`
Result on GCC/LLVM: `"hello"`
MSVC puts the whole `__VA_ARGS__` into the first arg of the called function-like macro.

This is basically a shortcoming in MSVC which can be solved by using argument `/Zc:preprocessor`.
So, we need to pass this argument to all HPy extensions but only for compiler `msvc`.

I used condition `self.compiler == 'msvc' or get_default_compiler() == 'msvc'` because if you specify `--compiler=something`, this will end up in `self.compiler`. If nothing is specifed, the default compiler will be used.

Do you have a better idea how to do that?

Reply:
How do others work around this? For instance, there is `cpython/Python/getargs.c` which uses varargs but does not special-case MSVC

Reply:
> How do others work around this?

I don't know. It depends on how exactly you use it.

> For instance, there is cpython/Python/getargs.c which uses varargs but does not special-case MSVC

I just grepped through the CPython code base for `__VA_ARGS__` and didn't find a single usage. I think what you were referring to are the C functions using varargs. But that's completely different.

Reply:
Ooops, I used the wrong search scope: Anyway, I still just found very few usages like this in `Objects/stringlib/fastsearch.h`
```
# define LOG(...) printf(__VA_ARGS__)
```
In this case, it doesn't matter since the `__VA_ARGS__` are just passed on as a whole.

Reply:
https://stackoverflow.com/questions/32399191/va-args-expansion-using-msvc